### PR TITLE
gh-92256: Improve Argument Clinic parser error messages

### PR DESCRIPTION
**Pull request by erlend-aasland**

### CODE REVIEW

**Comment by erlend-aasland** [comment]
(I'm not sure this needs a NEWS item, but I'm happy to add one.)

**Comment by erlend-aasland** [comment]
Details:

<details>
<summary>- Behaviour pre this change</summary>

Whitespace before end line (disallowed with confusing error message):

```console
$ git diff
diff --git a/Modules/_sqlite/cursor.c b/Modules/_sqlite/cursor.c
index 0e903ade5b..5c1f2d4d87 100644
--- a/Modules/_sqlite/cursor.c
+++ b/Modules/_sqlite/cursor.c
@@ -32,7 +32,7 @@
 /*[clinic input]
 module _sqlite3
 class _sqlite3.Cursor "pysqlite_Cursor *" "clinic_state()->CursorType"
-[clinic start generated code]*/
+ [clinic start generated code]*/
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=3c5b8115c5cf30f1]*/
 
 /*
$ python3.11  Tools/clinic/clinic.py -f Modules/_sqlite/cursor.c 
Error in file "Modules/_sqlite/cursor.c" on line 71:
Illegal function name: [clinic start generated code]*/
```

Whitespace after end line (allowed):

```console
$ git diff
diff --git a/Modules/_sqlite/cursor.c b/Modules/_sqlite/cursor.c
index 0e903ade5b..958891b995 100644
--- a/Modules/_sqlite/cursor.c
+++ b/Modules/_sqlite/cursor.c
@@ -32,7 +32,7 @@
 /*[clinic input]
 module _sqlite3
 class _sqlite3.Cursor "pysqlite_Cursor *" "clinic_state()->CursorType"
-[clinic start generated code]*/
+[clinic start generated code]*/ 
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=3c5b8115c5cf30f1]*/
 
 /*
$ !py
python3.11  Tools/clinic/clinic.py -f Modules/_sqlite/cursor.c 
$
```

Garbage after end line (fails badly):
```console
$ git diff
diff --git a/Modules/_sqlite/cursor.c b/Modules/_sqlite/cursor.c
index 0e903ade5b..8df423fc80 100644
--- a/Modules/_sqlite/cursor.c
+++ b/Modules/_sqlite/cursor.c
@@ -32,7 +32,7 @@
 /*[clinic input]
 module _sqlite3
 class _sqlite3.Cursor "pysqlite_Cursor *" "clinic_state()->CursorType"
-[clinic start generated code]*/
+[clinic start generated code]*/ // this won't work
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=3c5b8115c5cf30f1]*/
 
 /*
$ python3.11  Tools/clinic/clinic.py -f Modules/_sqlite/cursor.c 
Error in file "Modules/_sqlite/cursor.c" on line 71:
Exception raised during parsing:
Traceback (most recent call last):
  File "/Users/erlendaasland/src/cpython.git/Tools/clinic/clinic.py", line 2025, in parse
    parser.parse(block)
    ^^^^^^^^^^^^^^^^^^^
  File "/Users/erlendaasland/src/cpython.git/Tools/clinic/clinic.py", line 4176, in parse
    self.state(line)
    ^^^^^^^^^^^^^^^^
  File "/Users/erlendaasland/src/cpython.git/Tools/clinic/clinic.py", line 4217, in state_dsl_start
    fields = shlex.split(line)
             ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/shlex.py", line 315, in split
    return list(lex)
           ^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/shlex.py", line 300, in __next__
    token = self.get_token()
            ^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/shlex.py", line 109, in get_token
    raw = self.read_token()
          ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/shlex.py", line 191, in read_token
    raise ValueError("No closing quotation")
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: No closing quotation
```

</details>

<details>
<summary>- Behaviour post this change</summary>

Whitespace before end line (currently disallowed):

```console
$ git diff
diff --git a/Modules/_sqlite/cursor.c b/Modules/_sqlite/cursor.c
index 0e903ade5b..5c1f2d4d87 100644
--- a/Modules/_sqlite/cursor.c
+++ b/Modules/_sqlite/cursor.c
@@ -32,7 +32,7 @@
 /*[clinic input]
 module _sqlite3
 class _sqlite3.Cursor "pysqlite_Cursor *" "clinic_state()->CursorType"
-[clinic start generated code]*/
+ [clinic start generated code]*/
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=3c5b8115c5cf30f1]*/
 
 /*

$ python3.11  Tools/clinic/clinic.py -f Modules/_sqlite/cursor.c 
Error in file "Modules/_sqlite/cursor.c" on line 35:
Whitespace is not allowed before the stop line
```

Whitespace after end line (allowed):

```console
$ git diff
diff --git a/Modules/_sqlite/cursor.c b/Modules/_sqlite/cursor.c
index 0e903ade5b..8e4889f162 100644
--- a/Modules/_sqlite/cursor.c
+++ b/Modules/_sqlite/cursor.c
@@ -32,7 +32,7 @@
 /*[clinic input]
 module _sqlite3
 class _sqlite3.Cursor "pysqlite_Cursor *" "clinic_state()->CursorType"
-[clinic start generated code]*/
+[clinic start generated code]*/    
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=3c5b8115c5cf30f1]*/
 
 /*
$ python3.11  Tools/clinic/clinic.py -f Modules/_sqlite/cursor.c 
```

Garbage after end line (disallowed):
```console
$ git diff
diff --git a/Modules/_sqlite/cursor.c b/Modules/_sqlite/cursor.c
index 0e903ade5b..8df423fc80 100644
--- a/Modules/_sqlite/cursor.c
+++ b/Modules/_sqlite/cursor.c
@@ -32,7 +32,7 @@
 /*[clinic input]
 module _sqlite3
 class _sqlite3.Cursor "pysqlite_Cursor *" "clinic_state()->CursorType"
-[clinic start generated code]*/
+[clinic start generated code]*/ // this won't work
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=3c5b8115c5cf30f1]*/
 
 /*
$ python3.11  Tools/clinic/clinic.py -f Modules/_sqlite/cursor.c 
Error in file "Modules/_sqlite/cursor.c" on line 35:
Garbage after stop line: '// this won't work'
```

</details>

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -1628,10 +1628,16 @@ def parse_clinic_block(self, dsl_name):
...
-            if not line.startswith(stop_line):
-                return False
-            remainder = line[len(stop_line):]
-            return (not remainder) or remainder.isspace()
+            if line.startswith(stop_line):
+                remainder = line[len(stop_line):]
+                if remainder.isspace():
+                    return True
+                else:
+                    fail(f"Garbage after stop line: '{remainder.strip()}'")
```
Comment (vstinner) [inline Tools/clinic/clinic.py]: Maybe just show the raw string:

```suggestion
                    fail(f"Garbage after stop line: {remainder!r}")
```

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -1628,10 +1628,16 @@ def parse_clinic_block(self, dsl_name):
...
+            # gh-92256: don't allow incorrectly formatted stop lines
+            elif line.lstrip().startswith(stop_line):
+                fail("Whitespace is not allowed before the stop line")
```
Comment (vstinner) [inline Tools/clinic/clinic.py]: Maybe show the line with quotes to show these spaces?

```suggestion
                fail("Whitespace is not allowed before the stop line: {stop_line!r}")
```

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yes, that's a nice improvement.

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Yep, thanks!

**Review by vstinner — APPROVED** [review]
LGTM.

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -1628,10 +1628,16 @@ def parse_clinic_block(self, dsl_name):
         def is_stop_line(line):
             # make sure to recognize stop line even if it
             # doesn't end with EOL (it could be the very end of the file)
-            if not line.startswith(stop_line):
-                return False
-            remainder = line[len(stop_line):]
-            return (not remainder) or remainder.isspace()
+            if line.startswith(stop_line):
+                remainder = line[len(stop_line):]
+                if remainder.isspace():
```
Comment (serhiy-storchaka) [inline Tools/clinic/clinic.py]: And what if it is empty?

File: Tools/clinic/clinic.py [hunk 4]
```diff
@@ -1628,10 +1628,16 @@ def parse_clinic_block(self, dsl_name):
...
+                    return True
+                else:
+                    fail(f"Garbage after stop line: {remainder!r}")
+            # gh-92256: don't allow incorrectly formatted stop lines
+            elif line.lstrip().startswith(stop_line):
+                fail("Whitespace is not allowed before the stop line: {stop_line!r}")
+            return False
```
Comment (serhiy-storchaka) [inline Tools/clinic/clinic.py]: I would rewrite it in more symmetric way:
```suggestion
            if line.startswith(stop_line):
                remainder = line[len(stop_line):]
                if remainder and not remainder.isspace():
                    fail(f"Garbage after stop line: {remainder!r}")
                return True
            else:
                # gh-92256: don't allow incorrectly formatted stop lines
                if line.lstrip().startswith(stop_line):
                    fail("Whitespace is not allowed before the stop line: {stop_line!r}")
                return False
```

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: We could also drop the `else`, but I'm fine with this as well :)

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Thanks!

File: Tools/clinic/clinic.py (hunk 3 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Thanks for that heads-up. FTR: this was addressed in https://github.com/python/cpython/pull/92268#discussion_r864706240

File: Tools/clinic/clinic.py (hunk 4 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Fixup in 650392841091e62be25aafab7e43a031af6cdab0

**Review by serhiy-storchaka — APPROVED** [review]

File: Tools/clinic/clinic.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Note that the comment at the top of the function is exactly about this.

**Comment by erlend-aasland** [comment]
BTW, I guess we should backport this through 3.9.

**Comment by serhiy-storchaka** [comment]
It is an internal tool. No NEWS entry is needed. The change is invisible for users.

**Comment by erlend-aasland** [comment]
> It is an internal tool. No NEWS entry is needed. The change is invisible for users.

I'll remove the NEWS entry; thanks again.

**Comment by erlend-aasland** [comment]
Serhiy and Victor have approved this change. I'll land this in a day or two.

**Pull request merged**